File: app/src/main/java/com/example/myapplication/fragments/ProfileFragment.java
package com.example.myapplication.fragments;

import android.app.AlertDialog;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import com.example.myapplication.R;
import com.example.myapplication.auth.SignUpActivity;
import com.example.myapplication.models.ApiResponse;
import com.example.myapplication.models.User;
import com.example.myapplication.network.ApiClient;
import com.example.myapplication.profile.AddressActivity;
import com.example.myapplication.profile.MyItemsActivity;
import com.example.myapplication.profile.NotificationActivity;
import com.example.myapplication.profile.SavedItemsActivity;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.HashMap;
import java.util.Locale;
import java.util.Map;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class ProfileFragment extends Fragment {

    private TextView tvNama, tvLokasi, tvJoined;
    private int userId;
    private String currentLokasi = "";

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater,
                             @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_profile, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        tvNama   = view.findViewById(R.id.tv_nama);
        tvLokasi = view.findViewById(R.id.tv_lokasi);
        tvJoined = view.findViewById(R.id.tv_joined);

        SharedPreferences prefs = requireActivity()
                .getSharedPreferences("circlo_prefs", Context.MODE_PRIVATE);
        userId = prefs.getInt("user_id", 0);

        // Load profil dari API
        loadProfile();

        // Klik lokasi untuk edit
        tvLokasi.setOnClickListener(v -> showEditLokasiDialog());

        // Menu navigasi
        view.findViewById(R.id.menu_barang_saya).setOnClickListener(v ->
                startActivity(new Intent(getActivity(), MyItemsActivity.class)));

        view.findViewById(R.id.menu_barang_tersimpan).setOnClickListener(v ->
                startActivity(new Intent(getActivity(), SavedItemsActivity.class)));

        view.findViewById(R.id.menu_alamat).setOnClickListener(v ->
                startActivity(new Intent(getActivity(), AddressActivity.class)));

        view.findViewById(R.id.menu_notifikasi).setOnClickListener(v ->
                startActivity(new Intent(getActivity(), NotificationActivity.class)));

        // Logout
        view.findViewById(R.id.menu_logout).setOnClickListener(v -> {
            new AlertDialog.Builder(getContext())
                    .setTitle("Logout")
                    .setMessage("Yakin ingin keluar?")
                    .setPositiveButton("Logout", (dialog, which) -> {
                        requireActivity().getSharedPreferences("circlo_prefs", Context.MODE_PRIVATE)
                                .edit().clear().apply();
                        startActivity(new Intent(getActivity(), SignUpActivity.class));
                        requireActivity().finishAffinity();
                    })
                    .setNegativeButton("Batal", null)
                    .show();
        });
    }

    private void loadProfile() {
        ApiClient.getApiService().getProfile(userId).enqueue(new Callback<ApiResponse<User>>() {
            @Override
            public void onResponse(Call<ApiResponse<User>> call,
                                   Response<ApiResponse<User>> response) {
                if (response.isSuccessful() && response.body() != null
                        && response.body().isSuccess()) {
                    User user = response.body().getUser();

                    tvNama.setText(user.getUsername());

                    if (user.getLokasi() != null && !user.getLokasi().isEmpty()) {
                        tvLokasi.setText( user.getLokasi());
                        currentLokasi = user.getLokasi();
                    } else {
                        tvLokasi.setText(" Tap untuk tambah lokasi");
                    }

                    // Format tanggal bergabung
                    if (user.getCreatedAt() != null) {
                        try {
                            SimpleDateFormat inputFormat = new SimpleDateFormat(
                                    "yyyy-MM-dd HH:mm:ss", Locale.getDefault());
                            SimpleDateFormat outputFormat = new SimpleDateFormat(
                                    "MMMM yyyy", new Locale("id", "ID"));
                            Date date = inputFormat.parse(user.getCreatedAt());
                            tvJoined.setText("Bergabung " + outputFormat.format(date));
                        } catch (Exception e) {
                            tvJoined.setText("Bergabung " + user.getCreatedAt());
                        }
                    }

                } else {
                    Toast.makeText(getContext(),
                            "Response gagal: " + response.code(), Toast.LENGTH_LONG).show();
                }
            }

            @Override
            public void onFailure(Call<ApiResponse<User>> call, Throwable t) {
                Toast.makeText(getContext(),
                        "Error: " + t.getMessage(), Toast.LENGTH_LONG).show();
            }
        });
    }

    private void showEditLokasiDialog() {
        EditText etLokasi = new EditText(getContext());
        etLokasi.setHint("Contoh: Jakarta, Indonesia");
        etLokasi.setText(currentLokasi);
        etLokasi.setPadding(40, 20, 40, 20);

        new AlertDialog.Builder(getContext())
                .setTitle("Edit Lokasi")
                .setView(etLokasi)
                .setPositiveButton("Simpan", (dialog, which) -> {
                    String lokasi = etLokasi.getText().toString().trim();
                    if (!lokasi.isEmpty()) {
                        updateLokasi(lokasi);
                    }
                })
                .setNegativeButton("Batal", null)
                .show();
    }

    private void updateLokasi(String lokasi) {
        Map<String, Object> body = new HashMap<>();
        body.put("user_id", userId);
        body.put("lokasi", lokasi);

        ApiClient.getApiService().updateProfile(body).enqueue(new Callback<ApiResponse<User>>() {
            @Override
            public void onResponse(Call<ApiResponse<User>> call,
                                   Response<ApiResponse<User>> response) {
                if (response.isSuccessful() && response.body() != null
                        && response.body().isSuccess()) {
                    currentLokasi = lokasi;
                    tvLokasi.setText( lokasi);
                    Toast.makeText(getContext(),
                            "Lokasi berhasil diperbarui", Toast.LENGTH_SHORT).show();
                } else {
                    Toast.makeText(getContext(),
                            "Gagal memperbarui lokasi", Toast.LENGTH_SHORT).show();
                }
            }

            @Override
            public void onFailure(Call<ApiResponse<User>> call, Throwable t) {
                Toast.makeText(getContext(),
                        "Tidak dapat terhubung ke server", Toast.LENGTH_SHORT).show();
            }
        });
    }
}

File: app/src/main/java/com/example/myapplication/fragments/ProfileFragment.kt
package com.example.myapplication.fragments

import android.app.AlertDialog
import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import android.widget.ProgressBar
import android.widget.TextView
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.lifecycle.lifecycleScope
import com.example.myapplication.R
import com.example.myapplication.auth.LoginOrSignUpActivity
import com.example.myapplication.models.Profile
import com.example.myapplication.network.SupabaseClientProvider
import com.example.myapplication.profile.AddressActivity
import com.example.myapplication.profile.MyItemsActivity
import com.example.myapplication.profile.NotificationActivity
import com.example.myapplication.profile.SavedItemsActivity
import io.github.jan.supabase.auth.auth
import io.github.jan.supabase.postgrest.postgrest
import kotlinx.coroutines.launch

class ProfileFragment : Fragment() {

    // Target donasi untuk mencapai level berikutnya
    private val ECO_TARGET = 10

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?) =
        inflater.inflate(R.layout.fragment_profile, container, false)

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        val tvNama: TextView = view.findViewById(R.id.tv_nama)
        val tvLokasi: TextView = view.findViewById(R.id.tv_lokasi)
        val tvJoined: TextView = view.findViewById(R.id.tv_joined)
        val tvDiposting: TextView = view.findViewById(R.id.tv_diposting)
        val tvDiterima: TextView = view.findViewById(R.id.tv_diterima)
        val tvDibagikan: TextView = view.findViewById(R.id.tv_dibagikan)
        val progressEco: ProgressBar = view.findViewById(R.id.progress_eco)
        val tvDampakDesc: TextView = view.findViewById(R.id.tv_dampak_desc)
        val tvEcoTarget: TextView = view.findViewById(R.id.tv_eco_target)

        val userId = SupabaseClientProvider.client
            .auth.currentSessionOrNull()?.user?.id ?: return

        // Load profil dari Supabase
        lifecycleScope.launch {
            try {
                val profile = SupabaseClientProvider.client
                    .postgrest["profiles"]
                    .select { filter { eq("id", userId) } }
                    .decodeSingle<Profile>()

                tvNama.text = profile.username
                tvLokasi.text = profile.lokasi.ifEmpty { "Tap untuk tambah lokasi" }
                tvJoined.text = "Bergabung ${profile.createdAt.take(7)}"

                // Hitung jumlah barang yang diposting user
                val barangList = SupabaseClientProvider.client
                    .postgrest["barang"]
                    .select { filter { eq("user_id", userId) } }
                    .decodeList<com.example.myapplication.models.BarangItem>()

                val totalPosted = barangList.size
                val totalDonated = barangList.count { it.status == "diambil" }
                val totalReceived = profile.totalReceived

                tvDiposting.text = totalPosted.toString()
                tvDibagikan.text = totalDonated.toString()
                tvDiterima.text = totalReceived.toString()

                // ── ECO PROGRESS BAR ──────────────────────────────────────
                // Progress berdasarkan total barang yang berhasil didonasikan
                val progress = ((totalDonated % ECO_TARGET) * 100) / ECO_TARGET
                val level = totalDonated / ECO_TARGET + 1
                progressEco.progress = progress

                tvDampakDesc.text = when {
                    totalDonated == 0 -> "Mulai bagikan barangmu dan bantu kurangi sampah!"
                    totalDonated < 5  -> "Keren! $totalDonated barang sudah mendapat kehidupan baru 🌱"
                    totalDonated < 10 -> "Luar biasa! Kamu sudah menyelamatkan $totalDonated barang ♻️"
                    else              -> "Pahlawan lingkungan! $totalDonated barang sudah dibagikan 🌍"
                }

                val sisaKeDonasi = ECO_TARGET - (totalDonated % ECO_TARGET)
                tvEcoTarget.text = if (totalDonated == 0)
                    "Donasikan $ECO_TARGET barang untuk naik level!"
                else
                    "Level $level • $sisaKeDonasi donasi lagi untuk level ${level + 1}!"

            } catch (e: Exception) {
                Toast.makeText(context, "Error: ${e.message}", Toast.LENGTH_SHORT).show()
            }
        }

        // Klik lokasi → edit
        view.findViewById<TextView>(R.id.tv_lokasi).setOnClickListener {
            showEditLokasiDialog(userId)
        }

        // Menu navigasi
        view.findViewById<View>(R.id.menu_barang_saya)
            .setOnClickListener { startActivity(Intent(activity, MyItemsActivity::class.java)) }
        view.findViewById<View>(R.id.menu_barang_tersimpan)
            .setOnClickListener { startActivity(Intent(activity, SavedItemsActivity::class.java)) }
        view.findViewById<View>(R.id.menu_alamat)
            .setOnClickListener { startActivity(Intent(activity, AddressActivity::class.java)) }
        view.findViewById<View>(R.id.menu_notifikasi)
            .setOnClickListener { startActivity(Intent(activity, NotificationActivity::class.java)) }

        // Logout
        view.findViewById<View>(R.id.menu_logout).setOnClickListener {
            AlertDialog.Builder(context)
                .setTitle("Logout")
                .setMessage("Yakin ingin keluar?")
                .setPositiveButton("Logout") { _, _ ->
                    lifecycleScope.launch {
                        SupabaseClientProvider.client.auth.signOut()
                        startActivity(Intent(activity, LoginOrSignUpActivity::class.java))
                        activity?.finishAffinity()
                    }
                }
                .setNegativeButton("Batal", null)
                .show()
        }
    }

    private fun showEditLokasiDialog(userId: String) {
        val etLokasi = EditText(context).apply {
            hint = "Contoh: Jakarta, Indonesia"
            setPadding(40, 20, 40, 20)
        }
        AlertDialog.Builder(context)
            .setTitle("Edit Lokasi")
            .setView(etLokasi)
            .setPositiveButton("Simpan") { _, _ ->
                val lokasi = etLokasi.text.toString().trim()
                if (lokasi.isNotEmpty()) {
                    lifecycleScope.launch {
                        try {
                            SupabaseClientProvider.client.postgrest["profiles"]
                                .update({ set("lokasi", lokasi) }) {
                                    filter { eq("id", userId) }
                                }
                            view?.findViewById<TextView>(R.id.tv_lokasi)?.text = lokasi
                            Toast.makeText(context, "Lokasi diperbarui", Toast.LENGTH_SHORT).show()
                        } catch (e: Exception) {
                            Toast.makeText(context, "Gagal: ${e.message}", Toast.LENGTH_SHORT).show()
                        }
                    }
                }
            }
            .setNegativeButton("Batal", null)
            .show()
    }
}